content/posts/marimo.md

```diff
@@ -0,0 +1,271 @@
+latter's robustness and thorough documentation.
+
+marimo's original implementation has three main parts:
+
+1. a frontend
+2. a server
+3. a backend (or "runtime") with a kernel that runs Python code.
+
+All that was required were
+minimal changes to the marimo frontend and runtime, and replacing the server
+with a lightweight bridge.
+
+### Frontend
+
+The marimo frontend uses the same build when running with Pyodide or when
+running with a Python server. This helps maintain consistency between the two
+experiences and avoid feature drift. The code path splits based off some
+feature flags that downloads the marimo wheel from PyPI and initializes the
+single-threaded, Pyodide-compatible marimo kernel. Passing only a filename and
+a message callback function, we initialize the kernel as an async, never-ending
+Python process.
+
+We interact with this kernel through a lightweight RPC bridge that sends requests to
+the kernel and receives responses through the callback passed. This allows for
+an asynchronous-like feel to the execution. To support interruption of Python
+code, we simply followed the [Pyodide
+documentation](https://pyodide.org/en/stable/usage/keyboard-interrupts.html) on
+using its `SharedArrayBuffer`-based mechanism.
+
+For performance, we cache our assets and [run Pyodide in a web
+worker](https://pyodide.org/en/stable/usage/webworker.html) to avoid blocking
+the main browser thread. This creates some extra complexity with type-safety
+and mimicking blocking RPCs.
+
+We heavily leverage the Emscripten filesystem. Code is passed from the URL hash
+(for ease of sharing), to the main frame, to the worker, and finally to the
+Emscripten filesystem. Emscripten supports many filesystem implementations such
+as IndexDB which we use to persist the user's files. We are working on our own
+implementation of the Emscripten filesystem built on top of any S3-compatible
+bucket. This means you'll be able to list, read, and write files from your S3
+bucket just by interacting with ”os” filesystem all in Python.
+
+In order to make the experience smooth in any browser environment, we do our
+best to auto-install packages. Whenever a cell is run, we try to install it
+through Pyodide or otherwise fallback to micropip.
+
+### Runtime
+
+Similar to our frontend, our runtime uses the same
+implementation for Pyodide and CPython. We introduced just a few branches basd
+on [whether the kernel is running under
+Pyodide](https://pyodide.org/en/stable/usage/faq.html#how-to-detect-that-code-is-run-with-pyodide),
+since some Python features, such as shared memory, threading, and
+multiprocessing, are not yet available in Pyodide. As one example, we replaced
+multiprocessing queues with `asyncio` queues.
+
+We added a new entrypoint for creating the marimo kernel when running under
+Pyodide, which receives messages from the RPC bridge in an asynchronous
+control loop. This entrypoint also sets up stream objects (_e.g._, for standard
```

Comment:
most style guides say not to italicize `e.g.` and other common Latin abbreviations
```suggestion
control loop. This entrypoint also sets up stream objects (e.g., for standard
```